data_collector/data_collector.py

```diff
@@ -23,36 +35,197 @@ def collect_metrics(self):
+        output_path.parent.mkdir(parents=True, exist_ok=True)
+        np.save(output_path, final_array)
+        print(f"💾 Saved SAR array to {output_path}")
+
+        return pd.DataFrame([{
+            "image_count": len(normalized),
+            "array_shape": str(final_array.shape),
+            "saved_path": str(output_path)
+        }])
+
+    def get_image_list_from_gee(self, max_images=50):
+        collection = (
             ee.ImageCollection("COPERNICUS/S1_GRD")
             .filterBounds(self.roi)
-            .filterDate("2025-01-01", "2025-03-31")
+            .filterDate("2024-07-01", "2025-03-31")
             .filter(ee.Filter.eq("instrumentMode", "IW"))
             .select(["VV", "VH"])
         )
-        count = dataset.size().getInfo()
-        print(f"📸 Number of Sentinel-1 images found: {count}")
+        total = collection.size().getInfo()
+        print(f"📸 Total available: {total}")
+        images = collection.toList(min(total, max_images))
+        return [ee.Image(images.get(i)).clip(self.roi) for i in range(min(total, max_images))]
+
+    def fetch_numpy_array(self, image):
+        return geemap.ee_to_numpy(image, region=self.roi, bands=["VV", "VH"], scale=10)
+
+    def get_most_common_shape(self, shapes):
+        return Counter(shapes).most_common(1)[0][0]
+
+    def resize_to_target(self, arr, target_shape):
+        return resize(arr, target_shape, preserve_range=True, anti_aliasing=True).astype(np.float32)
+    # --- Sub-methods ---
+
+    def _process_image_collection(self, dataset, max_images=50):
+        images = dataset.toList(max_images)
+        raw_arrays = []
+        shape_counter = Counter()
+
+        for i in range(max_images):
+            image = ee.Image(images.get(i)).clip(self.roi)
+            try:
+                arr = self._download_image_array(image, i, max_images)
+                if arr is None:
+                    continue
+
+                arr = self._normalize_image_shape(arr)
+                shape_counter[arr.shape] += 1
+                raw_arrays.append((i, arr))
+
+            except Exception as e:
+                print(f"❌ Failed to process image {i}: {e}")
+
+        if not raw_arrays:
+            raise RuntimeError("❌ No valid images collected from GEE.")
+
+        # ✅ Determine most common shape
+        most_common_shape = shape_counter.most_common(1)[0][0]
+        print(f"📏 Most common shape: {most_common_shape}")
 
-        sar_image = dataset.median().clip(self.roi)
-        print("🛰️ Image metadata:", sar_image.getInfo())
-        return sar_image
+        # ✅ Normalize all arrays to most common shape
+        collected_arrays = []
+        for i, arr in raw_arrays:
+            if arr.shape != most_common_shape:
+                print(f"🔄 Resizing image {i} from {arr.shape} to {most_common_shape}")
+                arr = resize(arr, most_common_shape, preserve_range=True, anti_aliasing=True).astype(np.float32)
+            collected_arrays.append(arr)
```

Comment:
Consider refactoring as a list comprehension, which is more amenable to parallelization when scaling up and distributed processing when scaling wide.
```suggestion
            collected_arrays = [
        resize(arr, most_common_shape, preserve_range=True, anti_aliasing=True).astype(np.float32)
        if arr.shape != most_common_shape else arr
        for _, arr in raw_arrays
    ]
```